src/AspNetCore/Src/aspnetcoreapplication.cxx

```diff
@@ -0,0 +1,373 @@
+        pDataChunks,
+        nChunks,
+        fAsync,
+        fMoreData,
+        pfnCompletionCallback,
+        pvCompletionContext,
+        &dwBytesSent,
+        pfCompletionExpected);
+
+    return hr;
+}
+
+extern "C" __declspec(dllexport) HRESULT http_flush_response_bytes(
+    IHttpContext* pHttpContext,
+    PFN_ASYNC_COMPLETION pfnCompletionCallback,
+    void* pvCompletionContext,
+    BOOL* pfCompletionExpected)
+{
+    auto pHttpResponse = (IHttpResponse2*)pHttpContext->GetResponse();
+
+    BOOL fAsync = TRUE;
+    BOOL fMoreData = TRUE;
+    DWORD dwBytesSent;
+
+    HRESULT hr = pHttpResponse->Flush(
+        fAsync,
+        fMoreData,
+        pfnCompletionCallback,
+        pvCompletionContext,
+        &dwBytesSent,
+        pfCompletionExpected);
+    return hr;
+}
+
+// Thread execution callback
+static void ExecuteAspNetCoreProcess(LPVOID pContext)
+{
+    auto pApplication = (ASPNETCORE_APPLICATION*)pContext;
+
+    pApplication->ExecuteApplication();
+}
+
+ASPNETCORE_APPLICATION* ASPNETCORE_APPLICATION::s_Application = NULL;
+
+
+void ASPNETCORE_APPLICATION::SetCallbackHandles(PFN_REQUEST_HANDLER request_handler, PFN_SHUTDOWN_HANDLER shutdown_handler, void* pvRequstHandlerContext, void* pvShutdownHandlerContext)
+{
+    m_RequestHandler = request_handler;
+    m_RequstHandlerContext = pvRequstHandlerContext;
+    m_ShutdownHandler = shutdown_handler;
+    m_ShutdownHandlerContext = pvShutdownHandlerContext;
+
+    // Initialization complete
+    SetEvent(m_InitalizeEvent);
+}
+
+
+
+HRESULT ASPNETCORE_APPLICATION::Initialize(ASPNETCORE_CONFIG * pConfig)
+{
+    DWORD dwTimeout;
+    DWORD dwResult;
+
+    m_pConfiguration = pConfig;
+
+    m_InitalizeEvent = CreateEvent(
+        NULL,   // default security attributes
+        TRUE,   // manual reset event
+        FALSE,  // not set
+        NULL);  // name
+
+    if (m_InitalizeEvent == NULL)
+    {
+        return HRESULT_FROM_WIN32(GetLastError());
+    }
+
+    m_hThread = CreateThread(
+        NULL,       // default security attributes
+        0,          // default stack size
+        (LPTHREAD_START_ROUTINE)ExecuteAspNetCoreProcess,
+        this,       // thread function arguments
+        0,          // default creation flags
+        NULL);      // receive thread identifier
+
+    if (m_hThread == NULL)
+    {
+        return HRESULT_FROM_WIN32(GetLastError());
+    }
+
+    // If the debugger is attached, never timeout
+    if (IsDebuggerPresent())
+    {
+        dwTimeout = INFINITE;
+    }
+    else
+    {
+        dwTimeout = pConfig->QueryStartupTimeLimitInMS();
+    }
+
+    // What should the timeout be? (This is sorta hacky)
+    const HANDLE pHandles[2]{ m_hThread, m_InitalizeEvent };
+
+    // Wait on either the thread to complete or the event to be set
+    dwResult = WaitForMultipleObjects(2, pHandles, FALSE, dwTimeout);
+
+    // It all timed out
+    if (dwResult == WAIT_TIMEOUT)
+    {
+        return E_FAIL;
```

Comment:
pleas return HRESULT_FROM_WIN32(dwResult) instead of general E_FAIL. It will help diagnose the failure